utility/app.go

```diff
@@ -0,0 +1,431 @@
+package utility
+
+import (
+	"math"
+	"math/big"
+
+	"github.com/pokt-network/pocket/shared/crypto"
+	"github.com/pokt-network/pocket/shared/types"
+	utilTypes "github.com/pokt-network/pocket/utility/types"
+)
+
+func (u *UtilityContext) HandleMessageStakeApp(message *utilTypes.MessageStakeApp) types.Error {
+	publicKey, er := crypto.NewPublicKeyFromBytes(message.PublicKey)
+	if er != nil {
+		return types.ErrNewPublicKeyFromBytes(er)
+	}
+	// ensure above minimum stake
+	minStake, err := u.GetAppMinimumStake()
+	if err != nil {
+		return err
+	}
+	amount, err := types.StringToBigInt(message.Amount)
+	if err != nil {
+		return err
+	}
+	if types.BigIntLessThan(amount, minStake) {
+		return types.ErrMinimumStake()
+	}
+	// ensure signer has sufficient funding for the stake
+	signerAccountAmount, err := u.GetAccountAmount(message.Signer)
+	if err != nil {
+		return err
+	}
+	signerAccountAmount.Sub(signerAccountAmount, amount)
+	if signerAccountAmount.Sign() == -1 {
+		return types.ErrInsufficientAmountError()
+	}
+	maxChains, err := u.GetAppMaxChains()
+	if err != nil {
+		return err
+	}
+	// validate number of chains
+	if len(message.Chains) > maxChains {
+		return types.ErrMaxChains(maxChains)
+	}
+	// update account amount
+	if err := u.SetAccountAmount(message.Signer, signerAccountAmount); err != nil {
+		return err
+	}
+	// move funds from account to pool
+	if err := u.AddPoolAmount(utilTypes.AppStakePoolName, amount); err != nil {
+		return err
+	}
+	// calculate maximum relays from stake amount
+	maxRelays, err := u.CalculateAppRelays(message.Amount)
+	if err != nil {
+		return err
+	}
+	// ensure app doesn't already exist
+	exists, err := u.GetAppExists(publicKey.Address())
```

Comment:
What happens if the app already exists but we went ahead and staked? Should this check come before?